### PR TITLE
Add AnonymousProduct type

### PR DESCRIPTION
**Pull request by cdonovick**

### CODE REVIEW

File: hwtypes/adt.py [hunk 1]
```diff
@@ -77,6 +77,11 @@ def value_dict(self):
             d[k] = getattr(self, k)
         return MappingProxyType(d)
 
+
+class Product(AnonymousProduct, metaclass=ProductMeta):
+    def __repr__(self):
+        return f'{type(self).__name__}({", ".join(f"{k}={v}" for k,v in self.value_dict.items())})'
```
Comment (leonardt) [inline hwtypes/adt.py]: Does this need to be explicitly overridden to get the right name? Seems identical to the above implementation for `AnonymousProduct` (maybe I'm missing a subtle difference)

Comment (cdonovick) [reply hwtypes/adt.py]: No, it does not.  Will remove.

**Review by leonardt — APPROVED** [review]
LGTM, one minor comment.  I suspect magma may need to be updated to be consistent with this. The introduction to the callback methods may help magma reuse more of the base class logic (it can implement the special behavior in some of the callbacks, and use generic versions for the rest).

**Comment by cdonovick** [comment]
@leonardt if there are more hooks that would be useful for magma we should add them.

**Comment by leonardt** [comment]
I'll have to review the code, but there certainly might be some more.  Also similar ones for Bits would be good too.  I'll get back to you

**Comment by leonardt** [comment]
Testing this locally it doesn't seem to break magma, so I would say it's safe to merge when ready and we can update magma to use these callbacks as a separate step.

**Comment by cdonovick** [comment]
Sounds good.   I didn't think it should break magma as the interfaces you overload were unaffected.  Rerunning garnet flow when it passes I will merge.

**Comment by Kuree** [comment]
Unfortunately buildkite links the build with particular commit on the GarnetFlow. Since this PR is filed during the flow outage, rerunning the flow won't work. You can amend your last commit and force push to this PR to trigger a new build.

**Comment by rsetaluri** [comment]
I've been preferring `git commit --allow-empty -m "Trigger CI"` and a normal (non-force) push

**Pull request merged**